File: app/src/main/java/com/ramez/shopp/adapter/AddressAdapter.java
package com.ramez.shopp.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Dialogs.ConfirmDialog;
import com.ramez.shopp.Models.AddressModel;
import com.ramez.shopp.Models.MemberModel;
import com.ramez.shopp.R;
import com.ramez.shopp.databinding.RowAddressItemBinding;

import java.util.List;


public class AddressAdapter extends RecyclerView.Adapter<AddressAdapter.Holder> {

    private Context context;
    private List<AddressModel> addressModelList;
    private OnDeleteClicked onDeleteClicked;
    private OnContainerSelect onContainerSelect;
    private boolean canSelect;
    MemberModel user;
    private int defaultAddressId;

    public AddressAdapter(Context context, List<AddressModel> addressModelList, boolean canSelect, OnDeleteClicked onDeleteClicked, OnContainerSelect OnContainerSelect) {
        this.context = context;
        this.addressModelList = addressModelList;
        this.onDeleteClicked = onDeleteClicked;
        this.onContainerSelect = OnContainerSelect;
        this.canSelect = canSelect;
        user = UtilityApp.getUserData();
        defaultAddressId = user.lastSelectedAddress;
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        RowAddressItemBinding itemView = RowAddressItemBinding.inflate(LayoutInflater.from(context), viewGroup, false);
        return new Holder(itemView);
    }

    @Override
    public void onBindViewHolder(Holder holder, int position) {
        RowAddressItemBinding binding = holder.binding;
        AddressModel addressModel = addressModelList.get(position);

        if (defaultAddressId == addressModel.getId()) {
//            MemberModel user = UtilityApp.getUserData();
//            user.setLastSelectedAddress(addressModel.getId());
//            UtilityApp.setUserData(user);
//            if (UtilityApp.getUserData().lastSelectedAddress == addressModel.getId()) {
            holder.binding.rbSelectAddress.setChecked(true);
//            }
        } else {
            holder.binding.rbSelectAddress.setChecked(false);
        }

        binding.tvAddressMark.setText(context.getString(R.string.ph).concat(" " + addressModel.getMobileNumber()));
        binding.tvAddressNote.setText(addressModel.getFullAddress());
        binding.tvaAddressTitle.setText(addressModel.getName());


        binding.deleteAddressBut.setOnClickListener(v -> {

            onDeleteClicked.onDeleteClicked(addressModel, UtilityApp.getUserData().lastSelectedAddress == position, position);
//            deleteAddressId(addressModelList.get(position).getId(),position);


        });

        if (position == getItemCount() - 1) {
            holder.binding.divider.setVisibility(View.GONE);
        } else {
            holder.binding.divider.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        return addressModelList.size();
    }

    public void deleteAddressId(int addressId, int position) {
        ConfirmDialog.Click click = new ConfirmDialog.Click() {
            @Override
            public void click() {
                new DataFeacher(false, (obj, func, IsSuccess) -> {
                    if (func.equals(Constants.ERROR)) {
                        Toast.makeText(context, "" + context.getString(R.string.error_in_data), Toast.LENGTH_SHORT).show();
                    } else if (func.equals(Constants.FAIL)) {
                        Toast.makeText(context, "" + context.getString(R.string.fail_delete_address), Toast.LENGTH_SHORT).show();

                    } else if (func.equals(Constants.NO_CONNECTION)) {
                        Toast.makeText(context, "" + context.getString(R.string.no_internet_connection), Toast.LENGTH_SHORT).show();

                    } else {
                        if (IsSuccess) {
                            addressModelList.remove(position);
                            notifyDataSetChanged();
                            notifyItemRemoved(position);


                        } else {

                            Toast.makeText(context, "" + context.getString(R.string.fail_to_get_data), Toast.LENGTH_SHORT).show();
                        }
                    }

                }).deleteAddressHandle(addressId);
            }
        };

        new ConfirmDialog(context, context.getString(R.string.want_to_delete_address), R.string.ok, R.string.cancel_label, click, null, false);

    }

    public interface OnContainerSelect {
        void onContainerSelectSelected(AddressModel addressesDM, boolean makeDefault);
    }

    public interface OnDeleteClicked {
        void onDeleteClicked(AddressModel addressModel, boolean isChecked, int position);

    }

    public class Holder extends RecyclerView.ViewHolder {
        RowAddressItemBinding binding;

        public Holder(RowAddressItemBinding view) {
            super(view.getRoot());
            binding = view;

            binding.rbSelectAddress.setEnabled(!canSelect);

            binding.rbSelectAddress.setOnClickListener(v -> {
                AddressModel addressModel = addressModelList.get(getBindingAdapterPosition());
//                user = UtilityApp.getUserData();
                if (addressModel.getId() != defaultAddressId) {
                    defaultAddressId = addressModel.getId();
//                    user.setLastSelectedAddress(addressModel.getId());
//                    UtilityApp.setUserData(user);
                    notifyDataSetChanged();
                    onContainerSelect.onContainerSelectSelected(addressModel, true);
                }

            });

            binding.container.setOnClickListener(view1 -> {
                AddressModel addressModel = addressModelList.get(getBindingAdapterPosition());
                onContainerSelect.onContainerSelectSelected(addressModel, false);

            });

        }


    }
}

File: app/src/main/java/com/amour/shop/adapter/AddressAdapter.kt
package com.amour.shop.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.amour.shop.ApiHandler.DataFeacher
import com.amour.shop.ApiHandler.DataFetcherCallBack
import com.amour.shop.Dialogs.ConfirmDialog
import com.amour.shop.Models.AddressModel
import com.amour.shop.Models.MemberModel
import com.amour.shop.R
import com.amour.shop.classes.Constants
import com.amour.shop.classes.UtilityApp.getUserData
import com.amour.shop.databinding.RowAddressItemBinding


class AddressAdapter(
    private val context: Context,
    private val isDeleteVisible: Boolean,
    private val addressModelList: MutableList<AddressModel>?,
    private val canSelect: Boolean,
    private val onDeleteClicked: OnDeleteClicked,
    private val onContainerSelect: OnContainerSelect
) :
        RecyclerView.Adapter<AddressAdapter.Holder>() {
    var user: MemberModel? = getUserData()
    private var defaultAddressId: Int
    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): Holder {
        val itemView = RowAddressItemBinding.inflate(LayoutInflater.from(context), viewGroup, false)
        return Holder(itemView)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: Holder, position: Int) {
        val binding = holder.binding
        val addressModel = addressModelList?.get(position)
        holder.binding.rbSelectAddress.isChecked = defaultAddressId == addressModel?.id
        binding.tvAddressMark.text = context.getString(R.string.ph) + " " + addressModel?.mobileNumber
        binding.tvAddressNote.text = addressModel?.fullAddress
        binding.tvaAddressTitle.text = addressModel?.name
        if (isDeleteVisible) {
            holder.binding.deleteAddressBut.visibility = View.VISIBLE
        } else {
            holder.binding.deleteAddressBut.visibility = View.GONE
        }
        binding.deleteAddressBut.setOnClickListener { v ->
            onDeleteClicked.onDeleteClicked(
                addressModel,
                getUserData()!!.lastSelectedAddress == position,
                position
            )
        }
        if (position == itemCount - 1) {
            holder.binding.divider.visibility = View.GONE
        } else {
            holder.binding.divider.visibility = View.VISIBLE
        }
    }

    override fun getItemCount(): Int {
        return addressModelList?.size?:0
    }

    fun deleteAddressId(addressId: Int, position: Int) {
        val click: ConfirmDialog.Click = object : ConfirmDialog.Click() {
            override fun click() {
                DataFeacher(false,object :
                    DataFetcherCallBack {
                    @SuppressLint("NotifyDataSetChanged")
                    override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                        if (func == Constants.ERROR) {
                            Toast.makeText(
                                context,
                                "" + context.getString(R.string.error_in_data),
                                Toast.LENGTH_SHORT
                            ).show()
                        } else if (func == Constants.FAIL) {
                            Toast.makeText(
                                context,
                                "" + context.getString(R.string.fail_delete_address),
                                Toast.LENGTH_SHORT
                            ).show()
                        } else if (func == Constants.NO_CONNECTION) {
                            Toast.makeText(
                                context,
                                "" + context.getString(R.string.no_internet_connection),
                                Toast.LENGTH_SHORT
                            ).show()
                        } else {
                            if (IsSuccess) {
                                addressModelList?.removeAt(position)
                                notifyDataSetChanged()
                                notifyItemRemoved(position)
                            } else {
                                Toast.makeText(
                                    context,
                                    "" + context.getString(R.string.fail_to_get_data),
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }                    }

                }).deleteAddressHandle(addressId)
            }
        }
        ConfirmDialog(
            context,
            context.getString(R.string.want_to_delete_address),
            R.string.ok,
            R.string.cancel_label,
            click,
            null,
            false
        )
    }

    interface OnContainerSelect {
        fun onContainerSelectSelected(addressesDM: AddressModel?, makeDefault: Boolean)
    }

    interface OnDeleteClicked {
        fun onDeleteClicked(addressModel: AddressModel?, isChecked: Boolean, position: Int)
    }

    inner class Holder(var binding: RowAddressItemBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        init {
            binding.rbSelectAddress.isEnabled = !canSelect
            binding.rbSelectAddress.setOnClickListener { v ->
                val addressModel = addressModelList?.get(bindingAdapterPosition)
                //                user = UtilityApp.getUserData();
                if (addressModel?.id != defaultAddressId) {
                    defaultAddressId = addressModel?.id?:0
                    //                    user.setLastSelectedAddress(addressModel.getId());
//                    UtilityApp.setUserData(user);
                    notifyDataSetChanged()
                    onContainerSelect.onContainerSelectSelected(addressModel, true)
                }
            }
            binding.container.setOnClickListener { view1 ->
                val position = bindingAdapterPosition
                if (position >= 0 && position < (addressModelList?.size ?: 0)) {
                    val addressModel = addressModelList?.get(position)
                    onContainerSelect.onContainerSelectSelected(addressModel, false)
                }
            }
        }
    }

    init {
        defaultAddressId = user!!.lastSelectedAddress
    }
}